numpy/testing/utils.py

```diff
@@ -23,6 +23,25 @@
 else:
     from StringIO import StringIO
 
+try:
+    from threading import Lock
+    # Warning filtering is generally not threadsafe in python, this is also
+    # true for `catch_warnings` or `suppress_warinings`. In NumPy 1.12
+    # however, `assert_equal` and the array comparison asserts, use this
+    # to filter out some comparison warnings. Since removing this filter
+    # may also affect downstream projects and skimage (and possibly more)
+    # do parallel manual parallel testing using `assert_equal`, a quick fix
+    # seems to be to lock the less obvious threading trap. Ideally (in
+    # master this is the case), there should simply not be warning filter
+    # logic in the assert functions themself.
+    # While probably not perfectly safe in principle, it is sufficient
+    # in the case of skimage and probably most testing scenarios and the
+    # chance of deadlocks seems very unlikely.
+    _assert_comparison_lock = Lock()
+except ImportError:
```

Comment:
Though, seems like in fft and some other places, we do not guard against it, in docutils there is some guard using `dummy_threading`.